// babel-plugin-transform-class-properties/lib/index.js
"use strict";

exports.__esModule = true;

exports.default = function (_ref) {
  var t = _ref.types;
  var findBareSupers = {
    Super: function Super(path) {
      if (path.parentPath.isCallExpression({
        callee: path.node
      })) {
        this.push(path.parentPath);
      }
    }
  };
  var referenceVisitor = {
    TypeAnnotation: function TypeAnnotation(path) {
      path.skip();
    },
    ReferencedIdentifier: function ReferencedIdentifier(path) {
      if (this.scope.hasOwnBinding(path.node.name)) {
        this.collision = true;
        path.skip();
      }
    }
  };
  var buildObjectDefineProperty = (0, _babelTemplate2.default)("\n    Object.defineProperty(REF, KEY, {\n      configurable: true,\n      enumerable: true,\n      writable: true,\n      value: VALUE\n    });\n  ");

  var buildClassPropertySpec = function buildClassPropertySpec(ref, _ref2, scope) {
    var key = _ref2.key,
        value = _ref2.value,
        computed = _ref2.computed;
    return buildObjectDefineProperty({
      REF: ref,
      KEY: t.isIdentifier(key) && !computed ? t.stringLiteral(key.name) : key,
      VALUE: value ? value : scope.buildUndefinedNode()
    });
  };

  var buildClassPropertyLoose = function buildClassPropertyLoose(ref, _ref3, scope) {
    var key = _ref3.key,
        value = _ref3.value,
        computed = _ref3.computed;
    return t.expressionStatement(t.assignmentExpression("=", t.memberExpression(ref, key, computed || t.isLiteral(key)), value ? value : scope.buildUndefinedNode()));
  };

  return {
    inherits: _babelPluginSyntaxClassProperties2.default,
    visitor: {
      Class: function Class(path, state) {
        var buildClassProperty = state.opts.loose ? buildClassPropertyLoose : buildClassPropertySpec;
        var isDerived = !!path.node.superClass;
        var constructor = void 0;
        var props = [];
        var body = path.get("body");

        for (var _iterator = body.get("body"), _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
          var _ref4;

          if (_isArray) {
            if (_i >= _iterator.length) break;
            _ref4 = _iterator[_i++];
          } else {
            _i = _iterator.next();
            if (_i.done) break;
            _ref4 = _i.value;
          }

          var _path2 = _ref4;

          if (_path2.isClassProperty()) {
            props.push(_path2);
          } else if (_path2.isClassMethod({
            kind: "constructor"
          })) {
            constructor = _path2;
          }
        }

        if (!props.length) return;
        var nodes = [];
        var ref = void 0;

        if (path.isClassExpression() || !path.node.id) {
          (0, _babelHelperFunctionName2.default)(path);
          ref = path.scope.generateUidIdentifier("class");
        } else {
          ref = path.node.id;
        }

        var instanceBody = [];

        for (var _i2 = 0; _i2 < props.length; _i2++) {
          var prop = props[_i2];
          var propNode = prop.node;
          if (propNode.decorators && propNode.decorators.length > 0) continue;
          var isStatic = propNode.static;

          if (isStatic) {
            nodes.push(buildClassProperty(ref, propNode, path.scope));
          } else {
            instanceBody.push(buildClassProperty(t.thisExpression(), propNode, path.scope));
          }
        }

        if (instanceBody.length) {
          if (!constructor) {
            var newConstructor = t.classMethod("constructor", t.identifier("constructor"), [], t.blockStatement([]));

            if (isDerived) {
              newConstructor.params = [t.restElement(t.identifier("args"))];
              newConstructor.body.body.push(t.returnStatement(t.callExpression(t.super(), [t.spreadElement(t.identifier("args"))])));
            }

            var _body$unshiftContaine = body.unshiftContainer("body", newConstructor);

            constructor = _body$unshiftContaine[0];
          }

          var collisionState = {
            collision: false,
            scope: constructor.scope
          };

          for (var _i3 = 0; _i3 < props.length; _i3++) {
            var _prop = props[_i3];

            _prop.traverse(referenceVisitor, collisionState);

            if (collisionState.collision) break;
          }

          if (collisionState.collision) {
            var initialisePropsRef = path.scope.generateUidIdentifier("initialiseProps");
            nodes.push(t.variableDeclaration("var", [t.variableDeclarator(initialisePropsRef, t.functionExpression(null, [], t.blockStatement(instanceBody)))]));
            instanceBody = [t.expressionStatement(t.callExpression(t.memberExpression(initialisePropsRef, t.identifier("call")), [t.thisExpression()]))];
          }

          if (isDerived) {
            var bareSupers = [];
            constructor.traverse(findBareSupers, bareSupers);

            for (var _i4 = 0; _i4 < bareSupers.length; _i4++) {
              var bareSuper = bareSupers[_i4];
              bareSuper.insertAfter(instanceBody);
            }
          } else {
            constructor.get("body").unshiftContainer("body", instanceBody);
          }
        }

        for (var _i5 = 0; _i5 < props.length; _i5++) {
          var _prop2 = props[_i5];

          _prop2.remove();
        }

        if (!nodes.length) return;

        if (path.isClassExpression()) {
          path.scope.push({
            id: ref
          });
          path.replaceWith(t.assignmentExpression("=", ref, path.node));
        } else {
          if (!path.node.id) {
            path.node.id = ref;
          }

          if (path.parentPath.isExportDeclaration()) {
            path = path.parentPath;
          }
        }

        path.insertAfter(nodes);
      },
      ArrowFunctionExpression: function ArrowFunctionExpression(path) {
        var classExp = path.get("body");
        if (!classExp.isClassExpression()) return;
        var body = classExp.get("body");
        var members = body.get("body");

        if (members.some(function (member) {
          return member.isClassProperty();
        })) {
          path.ensureBlock();
        }
      }
    }
  };
};

var _babelHelperFunctionName = require("babel-helper-function-name");

var _babelHelperFunctionName2 = _interopRequireDefault(_babelHelperFunctionName);

var _babelTemplate = require("babel-template");

var _babelTemplate2 = _interopRequireDefault(_babelTemplate);

var _babelPluginSyntaxClassProperties = require("babel-plugin-syntax-class-properties");

var _babelPluginSyntaxClassProperties2 = _interopRequireDefault(_babelPluginSyntaxClassProperties);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
