// babel-plugin-transform-numeric-separator/lib/index.js
"use strict";

exports.__esModule = true;

exports.default = function (_ref) {
  var t = _ref.types;

  function replaceNumberArg(_ref2) {
    var node = _ref2.node;

    if (node.callee.name !== "Number") {
      return;
    }

    var arg = node.arguments[0];

    if (!t.isStringLiteral(arg)) {
      return;
    }

    arg.value = arg.value.replace(/_/g, "");
  }

  return {
    inherits: _babelPluginSyntaxNumericSeparator2.default,
    visitor: {
      CallExpression: replaceNumberArg,
      NewExpression: replaceNumberArg,
      NumericLiteral: function NumericLiteral(_ref3) {
        var node = _ref3.node;
        var extra = node.extra;

        if (extra && /_/.test(extra.raw)) {
          extra.raw = extra.raw.replace(/_/g, "");
        }
      }
    }
  };
};

var _babelPluginSyntaxNumericSeparator = require("babel-plugin-syntax-numeric-separator");

var _babelPluginSyntaxNumericSeparator2 = _interopRequireDefault(_babelPluginSyntaxNumericSeparator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
